central/externalbackups/plugins/s3compatible/s3compatible.go

```diff
@@ -0,0 +1,254 @@
+package s3compatible
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"path/filepath"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/config"
+	"github.com/aws/aws-sdk-go-v2/credentials"
+	"github.com/aws/aws-sdk-go-v2/feature/s3/manager"
+	"github.com/aws/aws-sdk-go-v2/service/s3"
+	s3Types "github.com/aws/aws-sdk-go-v2/service/s3/types"
+	"github.com/aws/smithy-go"
+	"github.com/pkg/errors"
+	"github.com/stackrox/rox/central/externalbackups/plugins"
+	"github.com/stackrox/rox/central/externalbackups/plugins/types"
+	"github.com/stackrox/rox/generated/storage"
+	"github.com/stackrox/rox/pkg/errorhelpers"
+	"github.com/stackrox/rox/pkg/httputil/proxy"
+	"github.com/stackrox/rox/pkg/logging"
+	"github.com/stackrox/rox/pkg/urlfmt"
+)
+
+const (
+	backupMaxTimeout               = 4 * time.Hour
+	testMaxTimeout                 = 5 * time.Second
+	initialConfigurationMaxTimeout = 5 * time.Minute
+	formatKey                      = "backup_2006-01-02T15:04:05.zip"
+)
+
+var (
+	log = logging.LoggerForModule()
+)
+
+func init() {
+	plugins.Add(types.S3CompatibleType, func(backup *storage.ExternalBackup) (types.ExternalBackup, error) {
+		return newS3Compatible(backup)
+	})
+}
+
+// s3Compatible plugin for the S3 compatible backup integration.
+// As the official AWS S3 is deprecating the path-style bucket addressing but
+// this style is still used in non-AWS S3 compatible providers, we decided to
+// implement a new plugin for the latter.
+// Having the two plugins allows for a clear separation between official AWS
+// and non-AWS features.
+// Having the S3 compatible plugin separate will also give us more freedom to
+// change to a different package if the aws-sdk decides to drop the path-style
+// option in the future.
+//
+// Additionally, this new S3 compatible plugin already uses the aws-sdk-v2
+// while the S3 plugin still uses v1.
+// This is to allow backwards compatibility for customers that are using the S3
+// backup integration with GCS buckets. This is not possible to do with v2
+// because GCS alters the Accept-Encoding header, which breaks the v2 request
+// signature. See:
+// https://github.com/aws/aws-sdk-go-v2/issues/1816
+// Tested here:
+// https://github.com/stackrox/stackrox/pull/11761
+// Using the S3 backup integration interoperability with GCS has been deprecated
+// in 4.5.
+// In 4.7 the S3 backup integration will migrate to aws-sdk-v2.
```

Comment:
```suggestion
// In 4.7 the S3 backup integration will migrate to aws-sdk-v2.
```
I would remove this line to avoid it becoming stale. We are planning to do this, but who knows what will happen.

Reply:
👍